src/main/java/org/elasticsearch/indices/recovery/RecoveryStatus.java

```diff
@@ -188,10 +188,13 @@ public void fail(RecoveryFailedException e, boolean sendShardFailure) {
     public void markAsDone() {
         if (finished.compareAndSet(false, true)) {
             assert tempFileNames.isEmpty() : "not all temporary files are renamed";
-            indexShard.postRecovery("peer recovery done");
-            // release the initial reference. recovery files will be cleaned as soon as ref count goes to zero, potentially now
-            decRef();
-            listener.onRecoveryDone(state());
+            try {
+                indexShard.postRecovery("peer recovery done");
+            } finally {
+                // release the initial reference. recovery files will be cleaned as soon as ref count goes to zero, potentially now
+                decRef();
+                listener.onRecoveryDone(state());
```

Comment:
we are already done with the recovery here failing doesn't help - we get a shardClosed exception since we are shutting down I think the try finally is just fine